### PR TITLE
[Work in Progress] Option to enable partition of graph after AD

### PR DESCRIPTION
**Pull request by jupvfranco**
This PR adds an option to partition the graph after AD, but before adding the optimizer nodes. Its motivation is to allow for better automatic partitions.
The partition algorithm is very similar to the one introduced in PR https://github.com/microsoft/onnxruntime/pull/5940.

Code to make this available to the ORTTrainer python API will come in a different PR.

## Assignment constraints
We cannot assign any node to any pipeline stage. There are few constraints that make an assignment valid. We describe them here and implement them in `VerifyAssignment`:

1. Every (pipeline) stage s is assigned at least one operator. 
2. Each operator is assigned to one and only one stage and all operators are assigned to a stage s (the stage id must be within the limits).
3. Given two operators `o1` and `o2`, such that `o1` produces tensor `t`, which is consumed by `o2`. 
     a. If both `o1` and `o2` are forward operators, then we require `stage_of(o1) <= stage_of(o2)`.
     b. If both `o1` and `o2` are backward operators, then we require `stage_of(o1) >= stage_of(o2)`.
     c. If `o1` is a forward operator and `o2` is a backward one, then we require `stage_of(o1) == stage_of(o2)`. This includes two cases: When `t` is the result of some activation function to be used by a backward operator, then we want both operators in the same device, or otherwise we would need extra work for copying `t` to the device of the consumer; and when `t` is an output of the forward computation to be used by the loss function (all these ops should live in the last device). 
4. Let `w` be the name of a parameter/weights tensor which is used by a single layer, and `sw` be the name of a weights tensor shared by multiple layers (e.g., the embedding tensor which are used by the embedding layer and the output classification layer of bert MLM), then we have:
    a. All the consumers of `w` must be assigned to the same stage.
    b. All the backward consumers of `sw` must be in the same stage of the first forward consumer of `sw` (i.e., the embedding layer in our example). 
5. Because we run partition after AD and before adding the optimizer nodes, we need to be careful with the placement of operators that produce tensors used by the optimizer. In particular, we require that for any parameter tensor named `w`, if AD creates an op `o` that generates `w_grad`, then we require `o` and the consumer of `w` to be assigned to the same stage.

## Testing
  * Added a few `PipelinePartitionAfterAD` unit tests with manual partition of small graphs, just to make sure that the partition code runs without problems. Needed to move `DistributedRunTestContext` from GradientGraphBuilderTest to training_session_test_utils.
  * Ideally, we would compare the result of a computation with and without partition. However, I have not found tests for that in ORT, yet. 
  *  I generated many assignment files for bert-tiny to be passed to the `--mapping_file` flag, and for each file, ran: 
```
mpirun -n $N $BUILD_DIR/onnxruntime_training_bert \
  --model_name $BERT_DATA_DIR/bert_models/nv/bert-tiny/bert-tiny-uncased_L_3_H_128_A_2_V_30528_S_512_Dp_0.1_optimized_layer_norm_opset12 \
  --train_data_dir $BERT_DATA_DIR/bert_data/512/books_wiki_en_corpus/train \
  --test_data_dir $BERT_DATA_DIR/bert_data/512/books_wiki_en_corpus/test \
  --train_batch_size 1 --mode train --display_loss_steps 1 --optimizer adam \
  --learning_rate 0.006 --gradient_accumulation_steps 1 --num_train_steps 3 \
  --warmup_ratio 0 --warmup_mode Linear \
  --pipeline_parallel_size $N --mapping_file $assignmentfile --partition_after_ad
```
We used N=[2,3,4]. 
  * Partitioning of forward graphs (as below) should still work.
```
mpirun -n 2 $BUILD_DIR/onnxruntime_training_bert \
  --model_name $BERT_DATA_DIR/bert_models/nv/bert-tiny/bert-tiny-uncased_L_3_H_128_A_2_V_30528_S_512_Dp_0.1_optimized_layer_norm_opset12 \
  --train_data_dir $BERT_DATA_DIR/bert_data/512/books_wiki_en_corpus/train \
  --test_data_dir $BERT_DATA_DIR/bert_data/512/books_wiki_en_corpus/test \
  --train_batch_size 1 --mode train --display_loss_steps 1 --optimizer adam \
  --learning_rate 0.006 --gradient_accumulation_steps 1 --num_train_steps 3 \
  --warmup_ratio 0 --warmup_mode Linear --pipeline_parallel_size 2 --cut_group_info 418
```
  * In order to test PP+DP and PP+HP, we ran the following commands. PP+DP+HP was not tested yet.  
```
mpirun -n 4 $BUILD_DIR/onnxruntime_training_bert \
  --model_name $BERT_DATA_DIR/bert_models/nv/bert-tiny/bert-tiny-uncased_L_3_H_128_A_2_V_30528_S_512_Dp_0.1_optimized_layer_norm_opset12 \
  --train_data_dir $BERT_DATA_DIR/bert_data/512/books_wiki_en_corpus/train \
  --test_data_dir $BERT_DATA_DIR/bert_data/512/books_wiki_en_corpus/test \
  --train_batch_size 1 --mode train --display_loss_steps 1 --optimizer adam \
  --learning_rate 0.006 --gradient_accumulation_steps 1 --num_train_steps 3 \
  --warmup_ratio 0 --warmup_mode Linear \
  --pipeline_parallel_size 2 --mapping_file <assignment file for pp=2> --partition_after_ad \
  --data_parallel_size 2 --use_nccl
```
and

```
mpirun -n 4 $BUILD_DIR/onnxruntime_training_bert \
  --model_name $BERT_DATA_DIR/bert_models/nv/bert-tiny/bert-tiny-uncased_L_3_H_128_A_2_V_30528_S_512_Dp_0.1_optimized_layer_norm_opset12 \
  --train_data_dir $BERT_DATA_DIR/bert_data/512/books_wiki_en_corpus/train \
  --test_data_dir $BERT_DATA_DIR/bert_data/512/books_wiki_en_corpus/test \
  --train_batch_size 1 --mode train --display_loss_steps 1 --optimizer adam \
  --learning_rate 0.006 --gradient_accumulation_steps 1 --num_train_steps 3 \
  --warmup_ratio 0 --warmup_mode Linear \
  --pipeline_parallel_size 2 --mapping_file <assignment file for pp=2> --partition_after_ad \
  --horizontal_parallel_size 2
```

  * I also did some testing with a different optimizer (`[adam|lamb]`) and with gradient accumulation (`--gradient_accumulation_steps 48 --num_train_steps 96`).
```
mpirun -n 4 $BUILD_DIR/onnxruntime_training_bert \
  --model_name $BERT_DATA_DIR/bert_models/bert-tiny/bert-tiny-uncased_L_3_H_128_A_2_V_30528_S_512_Dp_0.1_optimized_layer_norm_opset12 \
  --train_data_dir $BERT_DATA_DIR/bert_data/512/books_wiki_en_corpus/train  \
  --test_data_dir $BERT_DATA_DIR/bert_data/512/books_wiki_en_corpus/test \
  --train_batch_size 1 --mode train --display_loss_steps 1 --optimizer [adam|] \
  --learning_rate 0.006 --gradient_accumulation_steps 48 --num_train_steps 96 \
  --warmup_ratio 0 --warmup_mode Linear --pipeline_parallel_size 4 \
  --mapping_file <assignment file for pp=4> --partition_after_ad
```

All the tests above ran to completion. 

Note: `$BUILD_DIR` above refers to a Release build.

## Current Limitations

* While partitioning bert-large with these changes, we observed that a process was stuck on waiting for another one that went OOM. I'll give a repro instructions for this separately.
* Partition doesn't seem to work when using mixed precision. Currently, we pass as an input to ORT, a mapping between node names (or the names of their produced tensors) and pipeline stage ids. However, when mix precision is enabled, the same node will have different names in different mpi ranks. I believe there is some source of non-determinism in mixed precision code. I'll also provide repro instructions for this separately.

## Performance
We are currently working on performance measurements. **Results will appear here.**

### CODE REVIEW

**Comment by snnn** [comment]
Close inactive PR with merge conflicts
